README.md
# Compressed Sensing tutorial

> - In this repository are available the Python and Matlab files to replicate the experiments from my Bachelor's thesis: "Compressed Sensing for seismic acquisition".
>
> - Additionally, csDockerfile is a dockerfile constructed with the dependencies needed to replicate the Python routines. In order to build an image with it, you have to run the following command:
> ```
> docker build -f (path to csDockerfile) (path to context)
> ```
>
> - The Matlab functions needed to replicate the scripts are also available to download in the "Matlab_dependencies" folder. (once downloaded, you need to add to the path the folder "Matlab_dependencies").
>
> - Also, you can find the PDF version of my thesis.
>
> **IMPORTANT**: in order to get the files to work as expected, the paths to read the image (LOLA.jpg) and the SEGY file (gob_20200731_synthetic_shot-gather.sgy) need to be changed to your own.


ChangeLog
2021-09-17  Uwe Brauer  <oub@mat.ucm.es>

	* CS_seismic_full.m (Read the segy file and extract the size of
	the array (the path should): 21 purpose of this code is a bit opaque 

	* lola_CS.py (main): 21 some documentation, but what is the purpose of this file?

	* FFT1D.ipynb: 21 no documentation

	* CS1D_l1.ipynb: 21 no documentation

	* CS1D_greedy.ipynb: 21 no documentation



CS2D_l1.py
import numpy as np
import tkinter
import matplotlib
import random
import cvxpy as cvx
import scipy.fftpack as spfft
from pylbfgs import owlqn
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt

filename="../data/gob_20200731_synthetic_shot-gather.sgy"

### Read a stream of the segy file using ObsPy
from obspy.io.segy.core import _read_segy
st = _read_segy(filename,
                textual_header_encoding="EBCDIC",
                unpack_trace_headers=True)

l1    = st[0].stats.npts ### Get the number of samples
l2    = len(st) ### Get number of traces
fs    = st[0].stats.sampling_rate ### Get the sampling rate(samples by second)
delta = st[0].stats.delta ### Get the time sampling (fs = 1/delta ?)

### Construct an ndarray of all traces.
data = np.stack([t.data for t in st.traces]).T ### shape = (l1, l2)
fig=plt.figure(figsize=(10,10))
plt.subplot(1,1,1)
plt.imshow(data,cmap='seismic',interpolation="nearest",extent=(1,l2+1,fs*l1,0),vmin=-10000,vmax=10000,aspect="auto")
plt.xlabel("Trace")
plt.ylabel("Time(s)")
plt.title("Segy file")
plt.colorbar()
plt.show()

# Coefficient for the L1 norm of variables (see OWL-QN algorithm)
ORTHANTWISE_C = 5

def dct2(x):
    """Return 2D discrete cosine transform.
    """
    return spfft.dct(
        spfft.dct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)


def idct2(x):
    """Return inverse 2D discrete cosine transform.
    """
    return spfft.idct(
        spfft.idct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)

sample_size = 0.5
# create random sampling index vector
k = round(l2 * sample_size)
ri = np.random.choice(l2, k, replace=False) # random sample of indices
b = data[:,ri]
b = b.flat
b = np.expand_dims(b, axis=1)


#k=int(l2*sample_size)
#decimateIdxs = random.sample(range(0, l2), k)
### Generate a mask for quickly extracting subarrays and its complement
#decimateMask = np.ones(data.shape, dtype=bool)
#decimateMask[:,decimateIdxs] = False


#data_error = np.copy(data); ### copy by value; deep-copy not required.
### Drop the traces (set them to zero)
#data_error[:,decimateIdxs] = 0



def evaluate(x, g, step):
    """An in-memory evaluation callback."""

    # we want to return two things: 
    # (1) the norm squared of the residuals, sum((Ax-b).^2), and
    # (2) the gradient 2*A'(Ax-b)

    # expand x columns-first
    x2 = x.reshape((l2, l1)).T

    # Ax is just the inverse 2D dct of x2
    Ax2 = idct2(x2)

    # stack columns and extract samples
    Ax = Ax2.T.flat[ri].reshape(b.shape)

    # calculate the residual Ax-b and its 2-norm squared
    Axb = Ax - b
    fx = np.sum(np.power(Axb, 2))

    # project residual vector (k x 1) onto blank image (ny x nx)
    Axb2 = np.zeros(x2.shape)
    Axb2.T.flat[ri] = Axb # fill columns-first

    # A'(Ax-b) is just the 2D dct of Axb2
    AtAxb2 = 2 * dct2(Axb2)
    AtAxb = AtAxb2.T.reshape(x.shape) # stack columns

    # copy over the gradient vector
    np.copyto(g, AtAxb)

    return fx

# create images of mask (for visualization)
#Z = [np.zeros(data.shape, dtype='uint8')]
mask = np.zeros(data.shape)
#Xm = np.zeros(data.shape)
mask[:,ri] = data[:,ri]
# take random samples of image, store them in a vector b
b = data[:,ri]
b = b.flat
#b = b.astype(float)

# perform the L1 minimization in memory
Xat2 = owlqn(l2*l1, evaluate, None, ORTHANTWISE_C)

# transform the output back into the spatial domain
Xat = Xat2.reshape(l2, l1).T # stack columns
Xa = idct2(Xat)

cs_interp.py
#!/usr/bin/env python

import numpy as np
import tkinter
import matplotlib
import random
import cvxpy as cvx
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
from timeit import default_timer as timer
### IMP: User must run the docker image with the following display settings:
### docker run --net=host -e DISPLAY -v $HOME/.Xauthority:/root/.Xauthority:rw'

filename="../data/gob_20200731_synthetic_shot-gather.sgy"

### Read a stream of the segy file using ObsPy
from obspy.io.segy.core import _read_segy
st = _read_segy(filename,
                textual_header_encoding="EBCDIC",
                unpack_trace_headers=True)

l1    = st[0].stats.npts ### Get the number of samples
l2    = len(st) ### Get number of traces
fs    = st[0].stats.sampling_rate ### Get the sampling rate(samples by second)
delta = st[0].stats.delta ### Get the time sampling (fs = 1/delta ?)

### Construct an ndarray of all traces.
data = np.stack([t.data for t in st.traces]).T ### shape = (l1, l2)

#fig=plt.figure(figsize=(10,10))
#plt.subplot(1,1,1)
#plt.imshow(data,cmap='seismic',interpolation="nearest",extent=(1,l2+1,fs*l1,0),vmin=-10000,vmax=10000,aspect="auto")
#plt.xlabel("Trace")
#plt.ylabel("Time(s)")
#plt.title("Segy file")
#plt.colorbar()
#plt.show()

### Generate the list of traces to be dropped
#numero=int(l2*(1-0.8))
#decimateIdxs = random.sample(range(0, l2), numero)
decimateIdxs = np.arange(1,l2,2) #Best option to experience aliasing

### Generate a mask for quickly extracting subarrays and its complement
decimateMask = np.zeros(data.shape, dtype=bool)
decimateMask[:,decimateIdxs] = True

data_error = np.copy(data); ### copy by value; deep-copy not required.
### Drop the traces (set them to zero)
data_error[:,decimateIdxs] = 0

#fig=plt.figure(figsize=(10,10))
#plt.subplot(1,1,1)
#plt.imshow(data,cmap='Greys_r',interpolation="nearest",extent=(1,l2+1,fs*l1,0),vmin=-10000,vmax=10000,aspect="auto")
#plt.imshow(data_error,cmap='seismic',interpolation="nearest",extent=(1,l2+1,fs*l1,0),vmin=-10000,vmax=10000,aspect="auto")
#plt.xlabel("Trace")
#plt.ylabel("Time(s)")
#plt.title("Segy file with dropped traces")
#plt.colorbar()
#plt.show()

### plot the know matrix where data==data_error
#known = np.zeros((l1, l2))
#known[data==data_error] = 1
#fig=plt.figure(figsize=(20,10))
#plt.subplot(1,1,1)
#plt.imshow(known,cmap='seismic',interpolation="nearest",extent=(1,l2+1,fs*l1,0),aspect="auto")
#plt.xlabel("Trace")
#plt.ylabel("Time(s)")
#plt.title("Segy file")
#plt.colorbar()
#plt.show()

### COMPRESSED SENSING PROBLEM

y=data_error[:,::2] #captions

n=400
ri=np.random.choice(l2,n,replace=False)
Phi = np.identity(l1)
Theta = np.fft.ifft2(Phi, norm='ortho')
T = Theta[ri] #random Dirac delta functions for max. incoherence

### Convex l_1 min problem

alpha = cvx.Variable(l1)
objetivo = cvx.Minimize(cvx.norm(alpha,1))
rest = [T@alpha == y]
prob = cvx.Problem(objetivo, rest)
result = prob.solve(verbose=True)
Xat2 = np.array(alpha.value).squeeze()

lola_traces.py
import numpy as np
import scipy.fftpack as spfft
import scipy.ndimage as spimg
import matplotlib.pyplot as plt
from pylbfgs import owlqn
import imageio

def dct2(x):
    """Return 2D discrete cosine transform.
    """
    return spfft.dct(
        spfft.dct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)


def idct2(x):
    """Return inverse 2D discrete cosine transform.
    """
    return spfft.idct(
        spfft.idct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)

def progress(x, g, fx, xnorm, gnorm, step, k, ls):
    """Just display the current iteration.
    """
    print('Iteration {}'.format(k))
    return 0

_image_dims = None  # track target image dimensions here
_ri_vector = None  # reference the random sampling indices here
_b_vector = None  # reference the sampled vector b here
_A_matrix = None  # reference the dct matrix operator A here

def evaluate(X, g, step):
    """An in-memory evaluation callback.
    """

    # we want to return two things:
    # (1) the norm squared of the residuals, sum((Ax-b).^2), and
    # (2) the gradient 2*A'(Ax-b)

    # Ax is just the inverse dct of the spectral image
    X2 = X.reshape((_image_dims[0], _image_dims[1]))
    Ax2 = idct2(X2)
    #Ax2 = np.fft.ifft2(x2, norm = 'ortho').real
    # stack columns and extract samples
    Ax_c = Ax2[:,_ri_vector]
    Ax = np.ravel(Ax_c.T)

    # calculate the residual Ax-b and its 2-norm squared
    Axb = Ax - _b_vector
    fx = np.sum(np.power(Axb, 2))

    # project residual vector (k x 1) onto blank image (ny x nx)
    Axb2 = np.zeros(X2.shape)
    Ax_reshaped=np.reshape(Ax,(-1,len(_ri_vector)),order='F')
    Axb2[:,_ri_vector] = Ax_reshaped

    # A^t(Ax-b) is just the 2D dct of Axb2
    AtAxb2 = 2 * dct2(Axb2)
    #AtAxb2 = 2 * np.abs(np.fft.fft2(Ax, norm = 'ortho'))
    AtAxb = AtAxb2.reshape(X.shape)  # stack columns

    # copy over the gradient vector
    np.copyto(g, AtAxb)

    return fx

SCALE=0.1
SAMPLE=0.3
Xorig = imageio.imread('LOLA.jpg', as_gray=True, pilmode='L')
X = spimg.zoom(Xorig, SCALE)
ny, nx = X.shape
_image_dims = (ny, nx)
k2 = round(nx * SAMPLE)
r2 = np.random.choice(nx, k2, replace = False)
b2_c = X[:,r2]
b2 = np.ravel(b2_c.T)
_ri_vector = r2
_b_vector = np.expand_dims(b2, axis=1)

# perform the L1 minimization in memory
Xat2 = owlqn(nx*ny, evaluate, progress, 10)

    # transform the output back into the spatial domain
Xat = Xat2.reshape(nx, ny)  # stack columns
Xa = idct2(Xat)
#Xa = np.abs(np.fft.ifft2(Xat, norm = 'ortho')).real
    # create images of mask (for visualization)
mask = np.zeros(X.shape)
mask[:,r2] = 255
Xm = 255 * np.ones(X.shape)    
Xm[:,r2] = X[:,r2]

    # display the result
f, ax = plt.subplots(1, 3, figsize=(14, 4))
ax[0].imshow(X, cmap='gray', interpolation='none')
ax[0].set_axis_off()
ax[1].imshow(Xm, cmap='gray', interpolation='none')
ax[1].set_axis_off()
ax[2].imshow(Xa, cmap='gray', interpolation='none')
ax[2].set_axis_off()
plt.show()


CS_seismic_full.m
%%% Read the segy file and extract the size of the array (the path should
%%% be where the .sgy file was downloaded).
D = read_segy_file('gob_20200731_synthetic_shot-gather.sgy');
D=D.traces;
nt = size(D,1);
nr = size(D,2);
%%% Normalize the data
%D = log10(1+1e3*abs(D)).*sign(D);

%%% Visualize the original data
figure(1);
imagesc(D); colormap('redblue'); colorbar;
title('Original data'); caxis([-1000,1000]);
xlabel('Traces'); ylabel('Time samples')

%%% Seed the random number generator and the subsampling factor
rng(123456789);
p = 0.5;
%%% Create a random permutation of the receiver indices
idx = randperm(nr);
idx = sort(idx(1:round(nr*p)));
%%% Create the sampling operator (to drop the traces)
R = opKron(opRestriction(nr, idx), opDirac(nt));
%%% Take out the dropped traces
RD = R*D(:);
%%% Re-insert the traces w/ missing data
Dest_adj = reshape(R'*RD(:), nt, nr);

%%% Visualize the missing data
figure(2);
imagesc(Dest_adj); colormap('redblue'); colorbar;
title('Dropped traces'); caxis([-1000,1000]);
xlabel('Traces'); ylabel('Time samples')

%%% Create the sensing operator (we will create F, the DFT/Wavelet operator, and then
%%% take the transpose since it is an orthonormal basis
F = opDFT2(nt, nr);
%F = opWavelet2(nt,nr,'Daubechies');
%F = opWavelet2(nt,nr,'Haar');
%F2 = opCurvelet
%%% Create the caption vector 'y' using the operator RD
y = RD(:);
%%% Create the Theta operator by taking random samples of the inverse
%%% Fourier transform operator F^t
T = R*F';
%%% Set the options for the basis pursuit solver
options = spgSetParms('optTol', 5e-3, 'bpTol', 5e-3,...
    'iterations', 100, 'verbosity', 1);
%%% Now solve the \ell_1 min problem
xest = spg_bp(T, y, options);

%%% Transform the solution to the temporal domain w/ the original shape
dest = F'*xest;
Dest = reshape(dest, nt, nr);
%%% Determine the accuracy of the recover by computing the residual and the
%%% signal-to-noise ratio (SNR)
Ddiff = D - Dest;
SNR = -20*log10(norm(Ddiff(:))/norm(D(:)));

%%% Visualize the results
figure(3);
imagesc(real(Dest)); colormap('redblue'); colorbar;
title('Basis Pursuit Recovery'); caxis([-1000, 1000]);
xlabel('Traces'); ylabel('Time sample')

figure(4);
imagesc(real(Ddiff)); colormap('redblue'); colorbar;
title('Residual'); caxis([-1000, 1000]);
xlabel('Traces'); ylabel('Time sample')




CS_seismic_patches.m
%%% Read the segy file and extract the size of the array (the path should
%%% be where the .sgy file was downloaded).
D = read_segy_file('gob_20200731_synthetic_shot-gather.sgy');
D = D.traces;
%D = D(1:200,300:499);%Patch #1
%D = D(421:620,551:750);%Patch #2
%D = D(1421:1620,400:599);%Patch #3
D = D(1100:1299,1:200);%Patch #4
nt = size(D,1);
nr = size(D,2);
%%% Normalize the data
%D = log10(1+1e3*abs(D)).*sign(D);

%%% Visualize the original data
figure(1);
imagesc(D); colormap('redblue'); colorbar;
title('Original data'); caxis([-1000,1000]);
xlabel('Traces'); ylabel('Time samples')

%%% Seed the random number generator and the subsampling factor
rng(123456789);
p = 0.5;
%%% Create a random permutation of the receiver indices
idx = randperm(nr);
idx = sort(idx(1:round(nr*p)));
%%% Create the sampling operator (to drop the traces)
R = opKron(opRestriction(nr, idx), opDirac(nt));
%%% Take out the dropped traces
RD = R*D(:);
%%% Re-insert the traces w/ missing data
Dest_adj = reshape(R'*RD(:), nt, nr);

%%% Visualize the missing data
figure(2);
imagesc(Dest_adj); colormap('redblue'); colorbar;
title('Dropped traces'); caxis([-1000,1000]);
xlabel('Traces'); ylabel('Time samples')

%%% Create the sensing operator (we will create F, the DFT/Wavelet operator, and then
%%% take the transpose since it is an orthonormal basis
F = opDFT2(nt, nr);
%F = opWavelet2(nt,nr,'Daubechies');
%F = opWavelet2(nt,nr,'Haar');
%F2 = opCurvelet
%%% Create the caption vector 'y' using the operator RD
y = RD(:);
%%% Create the Theta operator by taking random samples of the inverse
%%% Fourier transform operator F^t
T = R*F';
%%% Set the options for the basis pursuit solver
options = spgSetParms('optTol', 5e-3, 'bpTol', 5e-3,...
    'iterations', 100, 'verbosity', 1);
%%% Now solve the \ell_1 min problem
xest = spg_bp(T, y, options);

%%% Transform the solution to the temporal domain w/ the original shape
dest = F'*xest;
Dest = reshape(dest, nt, nr);
%%% Determine the accuracy of the recover by computing the residual and the
%%% signal-to-noise ratio (SNR)
Ddiff = D - Dest;
SNR = -20*log10(norm(Ddiff(:))/norm(D(:)));

%%% Visualize the results
figure(3);
imagesc(real(Dest)); colormap('redblue'); colorbar;
title('Basis Pursuit Recovery'); caxis([-1000, 1000]);
xlabel('Traces'); ylabel('Time sample')

figure(4);
imagesc(real(Ddiff)); colormap('redblue'); colorbar;
title('Residual'); caxis([-1000, 1000]);
xlabel('Traces'); ylabel('Time sample')




lola_CS.py
#!/usr/bin/env python

import os
import numpy as np
import scipy.fftpack as spfft
import scipy.ndimage as spimg
import matplotlib.pyplot as plt
from pylbfgs import owlqn
import imageio


# Fraction to scale the original image
SCALE = 1.0

# Fraction of the scaled image to randomly sample
SAMPLE = 0.1

# Coeefficient for the L1 norm of variables (see OWL-QN algorithm)
ORTHANTWISE_C = 5


def dct2(x):
    """Return 2D discrete cosine transform.
    """
    return spfft.dct(
        spfft.dct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)


def idct2(x):
    """Return inverse 2D discrete cosine transform.
    """
    return spfft.idct(
        spfft.idct(x.T, norm='ortho', axis=0).T, norm='ortho', axis=0)


def progress(x, g, fx, xnorm, gnorm, step, k, ls):
    """Just display the current iteration.
    """
    print('Iteration {}'.format(k))
    return 0


_image_dims = None  # track target image dimensions here
_ri_vector = None  # reference the random sampling indices here
_b_vector = None  # reference the sampled vector b here


def evaluate(x, g, step):
    """An in-memory evaluation callback.
    """

    # we want to return two things:
    # (1) the norm squared of the residuals, sum((Ax-b).^2), and
    # (2) the gradient 2*A'(Ax-b)

    # expand x columns-first
    x2 = x.reshape((_image_dims[1], _image_dims[0])).T

    # Ax is just the inverse 2D dct of x2
    Ax2 = idct2(x2)

    # stack columns and extract samples
    Ax = Ax2.T.flat[_ri_vector].reshape(_b_vector.shape)

    # calculate the residual Ax-b and its 2-norm squared
    Axb = Ax - _b_vector
    fx = np.sum(np.power(Axb, 2))

    # project residual vector (k x 1) onto blank image (ny x nx)
    Axb2 = np.zeros(x2.shape)
    Axb2.T.flat[_ri_vector] = Axb  # fill columns-first

    # A'(Ax-b) is just the 2D dct of Axb2
    AtAxb2 = 2 * dct2(Axb2)
    AtAxb = AtAxb2.T.reshape(x.shape)  # stack columns

    # copy over the gradient vector
    np.copyto(g, AtAxb)

    return fx


_A_matrix = None  # reference the dct matrix operator A here


def main():

    global _b_vector, _A_matrix, _image_dims, _ri_vector

    # read image in grayscale, then downscale it
    Xorig = imageio.imread('LOLA.jpg', as_gray=True, pilmode='L')
    X = spimg.zoom(Xorig, SCALE)
    ny, nx = X.shape

    # take random samples of image, store them in a vector b
    k = round(nx * ny * SAMPLE)
    ri = np.random.choice(nx*ny, k, replace=False)  # random sample of indices
    b = X.T.flat[ri].astype(float)  # important: cast to 64 bit

    # This method evaluates the objective function sum((Ax-b).^2) and its
    # gradient without ever actually generating A (which can be massive)
    # Our ability to do this stems from our knowledge that Ax is just the
    # sampled idct2 of the spectral image (x in matrix form).
    # save image dims, sampling vector, and b vector and to global vars
    _image_dims = (ny, nx)
    _ri_vector = ri
    _b_vector = np.expand_dims(b, axis=1)
    # perform the L1 minimization in memory
    Xat2 = owlqn(nx*ny, evaluate, progress, ORTHANTWISE_C)

    # transform the output back into the spatial domain
    Xat = Xat2.reshape(nx, ny).T  # stack columns
    Xa = idct2(Xat)

    # create images of mask (for visualization)
    mask = np.zeros(X.shape)
    mask.T.flat[ri] = 255
    Xm = 255 * np.ones(X.shape)
    Xm.T.flat[ri] = X.T.flat[ri]

    # display the result
    f, ax = plt.subplots(1, 3, figsize=(14, 4))
    ax[0].imshow(X, cmap='gray', interpolation='none')
    ax[0].set_axis_off()
    ax[1].imshow(Xm, cmap='gray', interpolation='none')
    ax[1].set_axis_off()
    ax[2].imshow(Xa, cmap='gray', interpolation='none')
    ax[2].set_axis_off()
    plt.show()


if __name__ == '__main__':
    main()
